### DISCUSSION TITLE
Websockets closing early on ARM

### DISCUSSION THREAD

**Discussion by SelfhostedPro** [fastapi/fastapi#7316] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.
### Example
This is as close to minimal as I can get. I'm still a bit new to FastAPI so it's not good and I doubt it will work.

The following code runs fine on x86 but on arm64 devices I'm getting the following error:
```
WebSocket is already in CLOSING or CLOSED state.
```
More detailed logs as well as source code available here: 
logs: https://github.com/SelfhostedPro/Yacht/issues/115
SourceCode: https://github.com/SelfhostedPro/Yacht/blob/master/backend/api/routers/apps.py

```Python
from fastapi import APIRouter, Depends, HTTPException, WebSocket, status, Cookie

import aiodocker
import docker as sdocker

@router.websocket("/stats")
async def dashboard(websocket: WebSocket):
    auth_success = await websocket_auth(websocket=websocket)
    if auth_success:
        await websocket.accept()
        tasks = []
        async with aiodocker.Docker() as docker: 
            containers = []
            _containers = await docker.containers.list()
            for _app in _containers: 
                if _app._container['State'] == 'running':
                    containers.append(_app)
            for app in containers:
                _name = app._container['Names'][0][1:]
                container: DockerContainer = await docker.containers.get(_name)
                stats = container.stats(stream=True)
                tasks.append(process_container(_name, stats, websocket))
            await asyncio.gather(*tasks)
    else:
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)

async def process_container(name, stats, websocket):
    cpu_total = 0.0
    cpu_system = 0.0
    cpu_percent = 0.0
    async for line in stats:
        mem_current = line["memory_stats"]["usage"]
        mem_total = line["memory_stats"]["limit"]

        try:
            cpu_percent, cpu_system, cpu_total = await calculate_cpu_percent2(line, cpu_total, cpu_system)
        except KeyError as e:
            print("error while getting new CPU stats: %r, falling back")
            cpu_percent = await calculate_cpu_percent(line)

        full_stats = {
            "name": name,
            "time": line['read'],
            "cpu_percent": cpu_percent,
            "mem_current": mem_current,
            "mem_total": line["memory_stats"]["limit"],
            "mem_percent": (mem_current / mem_total) * 100.0,
        }
        try:
            await websocket.send_text(json.dumps(full_stats))
        except Exception as e:
            pass
```
### Description
In this code I'm getting stats from docker-py about containers running on the host system. In here there's some calculation to get things going which will sometimes report this error but it shouldn't close the websocket (doesn't on x86 devices).

### Environment
The fastapi version is `fastapi==0.60.2`
The frontend is VueJS
OS: Linux (arm64)
Python: Python 3.8.5

### Additional Context

On x86 I'll sometimes run into this error when a container is in a failing state but it will still keep the websocket open and send data as needed for the other containers. I'm not sure if I'm doing things the wrong way or if there's an underlying issue with websockets on arm devices.

**Comment by SelfhostedPro** [comment]
I'll come through and fix formatting on this question in a few minutes. Sorry about that.

**Comment by SelfhostedPro** [comment]
There you go. Sorry about the formatting again!

**Comment by SelfhostedPro** [comment, marked as answer]
Sorry, this looks like it was related to the docker arm client and I have resolved as it was an issue on my end.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue @SelfhostedPro ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
